File: leetCode/array/container-with-most-water.go
// https://leetcode.com/problems/container-with-most-water/
func minInt(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func maxArea(height []int) int {
    water := 0
    lenH := len(height)
    
    for i, j := 0, lenH - 1; i < lenH && i < j; {
        h := minInt(height[i], height[j])
        w := h * (j - i)
        //fmt.Printf("i: %v, j: %v, h: %v, w: %v\n", i, j, h, w)
        if w > water {
            water = w
            //fmt.Printf("water: %v\n", water)
        }
        
        if height[i] < height[j] {
            i++
        } else {
            j--
        }
    }
    return water
}

File: leetCode/two_pointers/container-with-most-water.py
# https://leetcode.com/problems/container-with-most-water/
class Solution:
    def maxArea(self, height: List[int]) -> int:
        i = 0
        k = len(height) - 1
        max_found = 0 
        while i < k:
            # print(f"i: {i}, k: {k}, h[i]: {height[i]}, h[k]: {height[k]}, max_found: {max_found}")
            area = min(height[i], height[k]) * (k - i) 
            max_found = max(area, max_found)
            if height[i] < height[k]:
                i += 1
            elif height[i] > height[k]:
                k -= 1
            else:
                i += 1
        
        return max_found
